File: backend/internal/api/prayers.go
package api

import (
	"context"
	"database/sql"

	"github.com/kelbwah/charis/backend/internal/database"
	"github.com/kelbwah/charis/backend/internal/routes"
	"github.com/labstack/echo/v4"
)

func InitPrayerRoutes(api *echo.Group, ctx context.Context, db *sql.DB, queries *database.Queries) {
	postPrayerRoutes(api, ctx, db, queries)
	getPrayerRoutes(api, ctx, db, queries)
	putPrayerRoutes(api, ctx, db, queries)
	deletePrayerRoutes(api, ctx, db, queries)
}

func postPrayerRoutes(api *echo.Group, ctx context.Context, db *sql.DB, queries *database.Queries) {
	api.POST("/prayers", func(c echo.Context) error {
		return routes.CreatePrayer(c, ctx, db, queries)
	})
}

func getPrayerRoutes(api *echo.Group, ctx context.Context, db *sql.DB, queries *database.Queries) {
	api.GET("/prayers/:id", func(c echo.Context) error {
		return routes.GetPrayerByID(c, ctx, db, queries)
	})

	api.GET("/prayers/user", func(c echo.Context) error {
		return routes.GetPrayersByUser(c, ctx, db, queries)
	})

	api.GET("/prayers", func(c echo.Context) error {
		return routes.GetAllPrayers(c, ctx, db, queries)
	})

	api.GET("/prayers/:id/count", func(c echo.Context) error {
		return routes.GetPrayerCountByID(c, ctx, db, queries)
	})
}

func putPrayerRoutes(api *echo.Group, ctx context.Context, db *sql.DB, queries *database.Queries) {
	api.PUT("/prayers/:id", func(c echo.Context) error {
		return routes.UpdatePrayer(c, ctx, db, queries)
	})
}

func deletePrayerRoutes(api *echo.Group, ctx context.Context, db *sql.DB, queries *database.Queries) {
	api.DELETE("/prayers/:id", func(c echo.Context) error {
		return routes.DeletePrayer(c, ctx, db, queries)
	})
}

File: frontend/services/prayers.ts
import api from "./api";
import { Prayer } from "./types";
import { AxiosResponse } from "axios";

export const createPrayer = (
  prayerData: Partial<Prayer>
): Promise<AxiosResponse<Prayer>> => {
  return api.post<Prayer>("/prayers", prayerData);
};

export const createPrayerResponse = (
  prayerId: string,
  status: string
): Promise<AxiosResponse<any>> => {
  return api.post<any>(`/prayers/${prayerId}/response?status=${status}`, {});
};

export const getPrayerResponsesByUser = (): Promise<
  AxiosResponse<Prayer[]>
> => {
  return api.get<any>(`/prayers/user/responses`);
};

export const getAllPrayersByUser = (): Promise<AxiosResponse<Prayer[]>> => {
  return api.get<Prayer[]>("/prayers/user");
};

export const getAllPrayers = (
  query: string = ""
): Promise<AxiosResponse<Prayer[]>> => {
  const url = query ? `/prayers?${query}` : "/prayers";
  return api.get<Prayer[]>(url);
};

export const getPrayerById = (id: string): Promise<AxiosResponse<Prayer>> => {
  return api.get<Prayer>(`/prayers/${id}`);
};

export const getPrayerCountById = (id: string): Promise<AxiosResponse<any>> => {
  return api.get<any>(`/prayers/${id}/count`);
};

export const updatePrayer = (
  id: string | null,
  prayerData: Partial<Prayer>
): Promise<AxiosResponse<Prayer>> => {
  return api.put<Prayer>(`/prayers/${id}`, prayerData);
};

export const updatePrayerResponse = (
  prayerId: string,
  status: string
): Promise<AxiosResponse<any>> => {
  return api.put<any>(`/prayers/${prayerId}/response?status=${status}`, {});
};

export const deletePrayer = (id: string): Promise<AxiosResponse<any>> => {
  return api.delete<any>(`/prayers/${id}`);
};
